packages/telemetry/telemetry.ts

```diff
@@ -246,3 +252,89 @@ export default class Telemetry {
     }
   }
 }
+
+export enum GitState {
+  NotInstalled, // The git cli was not found.
+  NoRepository, // Git is installed but no repository was found.
+  Ok, // Git is installed, a repository is present.
+}
+
+class GitProperties {
+  static async contributors(sinceDaysAgo: number, cwd?: PathLike): Promise<Array<string>> {
+    const unixTimeNow = Math.floor(Number(new Date()) / 1000);
+    const unixTimeAgo = unixTimeNow - sinceDaysAgo * 24 * 60 * 60;
+    try {
+      const { stdout } = await exec(
+        [
+          'git',
+          cwd && `-C ${cwd.toString()}`,
+          '--no-pager',
+          'log',
+          `--since=${unixTimeAgo}`,
+          '--format="%ae"',
+        ].join(' ')
+      );
+      return [
+        ...stdout
+          .trim()
+          .split('\n')
+          .reduce((acc, email) => {
+            acc.add(email);
+            return acc;
+          }, new Set<string>()),
+      ];
+    } catch {
+      return [];
+    }
+  }
+
+  static async state(cwd?: PathLike): Promise<GitState> {
+    const commandProcess = (await spawn('git', ['status'], {
+      shell: true,
+      cwd: cwd?.toString(),
+    })) as ChildProcess;
```

Comment:
I observed this `await spawn` to cause no other code to be executed.
```
~/src/forem$ (cd ../appmap-js/packages/cli; npm start openapi -- --output-file openapi.yml --openapi-title 'Forem API V1' --openapi-version '1.0.0' -d ~/src/forem)
```
![gitcontributors_await_spawn](https://user-images.githubusercontent.com/111290954/217649474-cc7cb429-bebe-40b9-b811-35f5fabb629e.png)

Changing `await spawn` to `spawn` along with changing `sendTelemetry` to `await sendTelemetry` lead to all sendTelemetry code to be executed.

![gitcontributors_no_await_spawn](https://user-images.githubusercontent.com/111290954/217649502-6c4aed00-9bfb-411b-b531-74c31815c578.png)

The change is in this [commit](https://github.com/getappmap/appmap-js/commit/46ad19dc144c3887aa44ff9bc3f185d12cdff9d7).  Maybe you can cherry pick it.

Reply:
I have a different [commit](https://github.com/getappmap/appmap-js/commit/d99c15b0d2ba783251ced46bb02f919620300830) that changes telemetry.ts to use CommandRunner, which we know works, and extends it to store the process exit code.  It passes the cli testcases but not all scanner testcases.  

The main issue with using CommandRunner in telemetry.ts is that it makes the scanner package not compile because the path used to load CommandRunner is relative to `cli`.  `telemetry.ts` is used in scanner through a symlink:
```
~/src/appmap-js/packages/scanner/src$ ls -l telemetry.ts 
lrwxrwxrwx 1 test test 28 Dec 12 15:54 telemetry.ts -> ../../telemetry/telemetry.ts
```

I think CommandRunner should be stored in package `telemetry`.

Reply:
Put CommandRunner in `packages/telemetry` in different [branch](https://github.com/getappmap/appmap-js/compare/main...sw/fix/include-git-contrib-count__version2?expand=1) and fixed typing errors that affected only [node 18](https://app.travis-ci.com/github/getappmap/appmap-js/builds/260601956).  Now fails the smoketests on [CI](https://app.travis-ci.com/github/getappmap/appmap-js/jobs/595855958#L4653).
```
+yarn run appmap index --appmap-dir ruby
/tmp/appmap-smoke.FYsw/.pnp.cjs:13668
      Error.captureStackTrace(firstError);
            ^
Error: Qualified path resolution failed: we looked for the following paths, but none could be accessed.
Source path: /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/src/cmds/agentInstaller/commandStruct
Not found: /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/src/cmds/agentInstaller/commandStruct
Not found: /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/src/cmds/agentInstaller/commandStruct.js
Not found: /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/src/cmds/agentInstaller/commandStruct.json
Not found: /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/src/cmds/agentInstaller/commandStruct.node
Require stack:
- /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/built/telemetry.js
- /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/built/fingerprint/fingerprintWatchCommand.js
- /tmp/appmap-smoke.FYsw/.yarn/cache/@appland-appmap-file-e9e091de9c-d07512d66f.zip/node_modules/@appland/appmap/built/cli.js
```

Reply:
This issue is now fixed in https://github.com/getappmap/appmap-js/pull/1026

Reply:
This is because spawn doesn't take a callback or return a promise.